src/mlpack/core/tree/rectangle_tree/rectangle_tree_impl.hpp

```diff
@@ -1449,6 +1449,26 @@ void RectangleTree<MetricType, StatisticType, MatType, SplitType, DescentType,
   {
     children[i] = NULL;
   }
+
+  // If we are the root, we need to restore the dataset pointer throughout
+  if (!hasParent)
+  {
+    std::stack<RectangleTree*> stack;
+    for (size_t i = 0; i < numChildren; ++i)
+    {
+      stack.push(children[i]);
+    }
+    while (!stack.empty())
+    {
+      RectangleTree* node = stack.top();
+      stack.pop();
+      node->dataset = dataset;
+      for (size_t i = 0; i < numChildren; ++i)
+      {
+        stack.push(children[i]);
+      }
+    }
+  }
```

Comment:
@rcurtin I have applied the similar principle in `rectangle_tree` I am not sure if it is working correctly, the tests do not show an error, but it goes in a loop for a long time. Tell me what do you think :+1:

Reply:
I did the same thing actually. :)  Here's a bugfix, in the inner loop:

```
for (size_t i = 0; i < node->numChildren; ++i)
{
  stack.push(node->children[i]);
}
```

:)